File: packages/xyd-opensdk-go/src/gowriter.ts
// A small string-emitting Go writer (no Go toolchain). Richer than opencli2go's
// golit — it tracks *aliased* imports (needed once a project has internal
// sub-packages like param/apijson/option/requestconfig) and assembles files.

const TAB = '\t';

/** Tracks the import paths a file needs, with optional aliases, grouped std vs ext. */
export class Imports {
  private paths = new Map<string, string | undefined>();

  /** Add an import path (optionally aliased). Returns the qualifier to use in code. */
  add(path: string, alias?: string): string {
    if (!this.paths.has(path)) this.paths.set(path, alias);
    return alias ?? this.paths.get(path) ?? lastSegment(path);
  }

  get size(): number {
    return this.paths.size;
  }

  render(): string {
    if (this.paths.size === 0) return '';
    const entries = [...this.paths.entries()];
    const isExt = (p: string) => p.split('/')[0].includes('.');
    const line = ([p, alias]: [string, string | undefined]) =>
      `${TAB}${alias ? `${alias} ` : ''}${JSON.stringify(p)}`;
    const std = entries.filter(([p]) => !isExt(p)).sort(byPath).map(line);
    const ext = entries.filter(([p]) => isExt(p)).sort(byPath).map(line);
    const groups = [std, ext].filter((g) => g.length > 0);
    return `import (\n${groups.map((g) => g.join('\n')).join('\n\n')}\n)`;
  }
}

function lastSegment(path: string): string {
  const seg = path.split('/').pop() || path;
  return seg;
}

function byPath(a: [string, unknown], b: [string, unknown]): number {
  return a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0;
}

/** Assemble a complete Go file: package clause + imports + declarations. */
export function goFile(pkg: string, imports: Imports, decls: string[]): string {
  const parts = [`package ${pkg}`];
  if (imports.size > 0) parts.push(imports.render());
  parts.push(...decls.filter(Boolean));
  return `${parts.join('\n\n')}\n`;
}

/** A Go doc comment block (`// ...`) from free text, or '' when empty. */
export function goDoc(text: string | undefined, symbol?: string): string {
  const raw = (text || '').trim();
  if (!raw && !symbol) return '';
  const body = raw || `${symbol} ...`;
  return body
    .split('\n')
    .map((l) => (l.trim() ? `// ${l.trim()}` : '//'))
    .join('\n');
}

/** A Go struct declaration with pre-rendered `<name> <type> <tag?>` field lines. */
export function goStruct(name: string, fields: string[], doc = ''): string {
  const head = doc ? `${doc}\n` : '';
  if (fields.length === 0) return `${head}type ${name} struct{}`;
  const body = fields.map((f) => `${TAB}${f}`).join('\n');
  return `${head}type ${name} struct {\n${body}\n}`;
}

/** A raw Go struct field line: `Name Type \`tag\``. */
export function goField(name: string, type: string, tag?: string): string {
  return tag ? `${name} ${type} ${'`'}${tag}${'`'}` : `${name} ${type}`;
}

File: crates/xyd_opensdk_go/src/gowriter.rs
//! The string-emitting Go writer — port of gowriter.ts (Imports + goFile/goDoc/
//! goStruct/goField).

const TAB: &str = "\t";

/// Tracks import paths (insertion-ordered), with optional aliases.
#[derive(Default)]
pub struct Imports {
    // (path, alias) in insertion order; first-write wins (Map.has guard).
    paths: Vec<(String, Option<String>)>,
}

impl Imports {
    pub fn new() -> Self {
        Self::default()
    }

    /// Add an import path (optionally aliased). Returns the qualifier to use.
    pub fn add(&mut self, path: &str, alias: Option<&str>) -> String {
        if let Some((_, existing)) = self.paths.iter().find(|(p, _)| p == path) {
            // returns alias ?? existing ?? lastSegment(path)
            return alias
                .map(|a| a.to_string())
                .or_else(|| existing.clone())
                .unwrap_or_else(|| last_segment(path));
        }
        self.paths
            .push((path.to_string(), alias.map(|a| a.to_string())));
        alias
            .map(|a| a.to_string())
            .unwrap_or_else(|| last_segment(path))
    }

    pub fn size(&self) -> usize {
        self.paths.len()
    }

    pub fn render(&self) -> String {
        if self.paths.is_empty() {
            return String::new();
        }
        let is_ext = |p: &str| p.split('/').next().unwrap_or("").contains('.');
        let line = |(p, alias): &(String, Option<String>)| -> String {
            format!(
                "{TAB}{}{}",
                alias.as_ref().map(|a| format!("{a} ")).unwrap_or_default(),
                crate::naming::json_string(p)
            )
        };
        let mut std: Vec<&(String, Option<String>)> =
            self.paths.iter().filter(|(p, _)| !is_ext(p)).collect();
        let mut ext: Vec<&(String, Option<String>)> =
            self.paths.iter().filter(|(p, _)| is_ext(p)).collect();
        std.sort_by(|a, b| a.0.cmp(&b.0));
        ext.sort_by(|a, b| a.0.cmp(&b.0));
        let std_lines: Vec<String> = std.iter().map(|e| line(e)).collect();
        let ext_lines: Vec<String> = ext.iter().map(|e| line(e)).collect();
        let groups: Vec<String> = [std_lines, ext_lines]
            .into_iter()
            .filter(|g| !g.is_empty())
            .map(|g| g.join("\n"))
            .collect();
        format!("import (\n{}\n)", groups.join("\n\n"))
    }
}

fn last_segment(path: &str) -> String {
    path.rsplit('/').next().unwrap_or(path).to_string()
}

pub fn go_file(pkg: &str, imports: &Imports, decls: &[String]) -> String {
    let mut parts: Vec<String> = vec![format!("package {pkg}")];
    if imports.size() > 0 {
        parts.push(imports.render());
    }
    for d in decls {
        if !d.is_empty() {
            parts.push(d.clone());
        }
    }
    format!("{}\n", parts.join("\n\n"))
}

pub fn go_doc(text: Option<&str>, symbol: Option<&str>) -> String {
    let raw = text.unwrap_or("").trim();
    if raw.is_empty() && symbol.is_none() {
        return String::new();
    }
    let body = if raw.is_empty() {
        format!("{} ...", symbol.unwrap())
    } else {
        raw.to_string()
    };
    body.split('\n')
        .map(|l| {
            if l.trim().is_empty() {
                "//".to_string()
            } else {
                format!("// {}", l.trim())
            }
        })
        .collect::<Vec<_>>()
        .join("\n")
}

pub fn go_struct(name: &str, fields: &[String], doc: &str) -> String {
    let head = if doc.is_empty() {
        String::new()
    } else {
        format!("{doc}\n")
    };
    if fields.is_empty() {
        return format!("{head}type {name} struct{{}}");
    }
    let body = fields
        .iter()
        .map(|f| format!("{TAB}{f}"))
        .collect::<Vec<_>>()
        .join("\n");
    format!("{head}type {name} struct {{\n{body}\n}}")
}

pub fn go_field(name: &str, ty: &str, tag: Option<&str>) -> String {
    match tag {
        Some(t) => format!("{name} {ty} `{t}`"),
        None => format!("{name} {ty}"),
    }
}
